### PR TITLE
Simplify workflow

### PR DESCRIPTION
**Pull request by eivindjahren**
**Issue**
work towards #2885 

**Approach**
_Short description of the approach_

## Pre review checklist

- [ ] Added appropriate release note label
- [ ] PR title captures the intent of the changes, and is fitting for release notes.
- [ ] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).

Adding labels helps the maintainers when writing release notes. This is the [list of release note labels](https://github.com/equinor/ert/labels?q=release-notes).

### CODE REVIEW

File: src/ert/_c_wrappers/util/substitution_list.py [hunk 1]
```diff
@@ -115,6 +117,14 @@ def indexForKey(self, key):
     def substitute(self, to_substitute: str) -> str:
         return self._alloc_filtered_string(to_substitute)
 
+    def substitute_file(self, to_substitute: str) -> str:
+        tmpdir = mkdtemp("ert_workflow")
+        tmp_file = os.path.join(tmpdir, "ert-workflow")
+        if self._filter_file(to_substitute, tmp_file):
+            return tmp_file
+        else:
+            return to_substitute
```
Comment (valentin-krasontovitsch) [inline src/ert/_c_wrappers/util/substitution_list.py:126]: please change this. : )

**Review by valentin-krasontovitsch — CHANGES_REQUESTED** [review]
managed to do a proper read through after all - it looks good, and i do have a few comments / suggestions.

File: src/ert/_c_wrappers/util/substitution_list.py (hunk 1 above)
Comment (eivindjahren) [reply src/ert/_c_wrappers/util/substitution_list.py]: Done.

File: src/ert/_c_wrappers/job_queue/workflow.py [hunk 2]
```diff
@@ -16,41 +12,63 @@
...
-        args = self._iget_args(index)
-        return job, args
+        return self.cmd_list[index]
 
     def __iter__(self):
-        for index in range(len(self)):
-            yield self[index]
+        return iter(self.cmd_list)
+
+    def _try_compile(self, context: Optional["SubstitutionList"]):
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply src/ert/_c_wrappers/job_queue/workflow.py]: Changed it to parse :)

File: src/ert/_c_wrappers/job_queue/workflow.py [hunk 3]
```diff
@@ -16,41 +12,63 @@
...
+        self.cmd_list = []
+        parser = self.job_list.parser
+        content = parser.parse(to_compile)
+
+        errors = content.getErrors()
+        if errors:
+            return errors
+
+        for line in content:
+            print(list(line))
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply src/ert/_c_wrappers/job_queue/workflow.py]: Removed

File: src/ert/_c_wrappers/job_queue/workflow.py [hunk 4]
```diff
@@ -16,41 +12,63 @@
...
+
+        self._try_compile(None)
 
     def __len__(self):
-        return self._count()
+        return len(self.cmd_list)
+
+    @property
+    def compiled(self):
+        return self._compile_time is None
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply src/ert/_c_wrappers/job_queue/workflow.py]: Removed the whole concept of mtime etc.

File: src/ert/_c_wrappers/util/substitution_list.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (eivindjahren) [reply src/ert/_c_wrappers/util/substitution_list.py]: We could do a `try: ... finally:` I am not worried about running out of tmpdirs though :)

File: src/ert/_c_wrappers/job_queue/workflow.py [hunk 5]
```diff
@@ -16,41 +12,63 @@
...
+    def _try_compile(self, context: Optional["SubstitutionList"]):
+        to_compile = self.src_file
+        if not os.path.exists(self.src_file):
+            raise ValueError(f"Workflow file {self.src_file} does not exist")
+        if context is not None:
+            to_compile = context.substitute_file(self.src_file)
+            self._compile_time = None
+        mtime = os.path.getmtime(to_compile)
+        if self._compile_time is not None and mtime <= self._compile_time:
+            return
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply src/ert/_c_wrappers/job_queue/workflow.py]: I removed the entire concept of trying to cache the compile, I am not convinced it ever happens. So I just made it recompile every time.

File: src/ert/_c_wrappers/util/substitution_list.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/util/substitution_list.py]: great then let's just keep it simple and not delete them : )

**Review by valentin-krasontovitsch — APPROVED** [review]
besides the question about logging, this looks good : )

File: src/ert/_c_wrappers/job_queue/workflow.py [hunk 6]
```diff
@@ -73,13 +73,14 @@ def run(
...
-        if errors is not None:
+        try:
+            self._parse_src_file(context)
+        except ValueError as err:
             msg = (
                 "** Warning: The workflow file {} is not valid - "
                 "make sure the workflow jobs are defined correctly:\n {}\n"
             )
-            sys.stderr.write(msg.format(self.src_file, errors))
+            sys.stderr.write(msg.format(self.src_file, err))
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply src/ert/_c_wrappers/job_queue/workflow.py:83]: It unfortunately does not as warnings are ignored by default.

[the remark this answers is not in the document]
Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/job_queue/workflow.py:83]: ok, I see, but don't we also want to log these using the logger?

**Pull request merged**